server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationState.java

```diff
@@ -0,0 +1,485 @@
+        if (startJoinRequest.getTerm() <= getCurrentTerm()) {
+            logger.debug("handleStartJoin: ignored as term provided [{}] not greater than current term [{}]",
+                startJoinRequest.getTerm(), getCurrentTerm());
+            throw new CoordinationStateRejectedException("incoming term " + startJoinRequest.getTerm() +
+                " not greater than current term " + getCurrentTerm());
+        }
+
+        logger.debug("handleStartJoin: updating term from [{}] to [{}]", getCurrentTerm(), startJoinRequest.getTerm());
+
+        persistedState.setCurrentTerm(startJoinRequest.getTerm());
+        assert getCurrentTerm() == startJoinRequest.getTerm();
+        lastPublishedVersion = 0;
+        lastPublishedConfiguration = getLastAcceptedConfiguration();
+        startedJoinSinceLastReboot = true;
+        electionWon = false;
+        joinVotes = new VoteCollection();
+        publishVotes = new VoteCollection();
+
+        return new Join(localNode, startJoinRequest.getSourceNode(), getCurrentTerm(), getLastAcceptedTerm(), getLastAcceptedVersion());
+    }
+
+    /**
+     * May be called on receipt of a Join.
+     *
+     * @param join The Join received.
+     * @return true iff this instance does not already have a join vote from the given source node for this term
+     * @throws CoordinationStateRejectedException if the arguments were incompatible with the current state of this object.
+     */
+    public boolean handleJoin(Join join) {
+        assert join.getTargetNode().equals(localNode) : "handling join " + join + " for the wrong node " + localNode;
+
+        if (join.getTerm() != getCurrentTerm()) {
+            logger.debug("handleJoin: ignored join due to term mismatch (expected: [{}], actual: [{}])",
+                getCurrentTerm(), join.getTerm());
+            throw new CoordinationStateRejectedException(
+                "incoming term " + join.getTerm() + " does not match current term " + getCurrentTerm());
+        }
+
+        if (startedJoinSinceLastReboot == false) {
+            logger.debug("handleJoin: ignored join as term was not incremented yet after reboot");
+            throw new CoordinationStateRejectedException("ignored join as term has not been incremented yet after reboot");
+        }
+
+        final long lastAcceptedTerm = getLastAcceptedTerm();
+        if (join.getLastAcceptedTerm() > lastAcceptedTerm) {
+            logger.debug("handleJoin: ignored join as joiner has a better last accepted term (expected: <=[{}], actual: [{}])",
+                lastAcceptedTerm, join.getLastAcceptedTerm());
+            throw new CoordinationStateRejectedException("incoming last accepted term " + join.getLastAcceptedTerm() +
+                " of join higher than current last accepted term " + lastAcceptedTerm);
+        }
+
+        if (join.getLastAcceptedTerm() == lastAcceptedTerm && join.getLastAcceptedVersion() > getLastAcceptedVersion()) {
```

Comment:
Tests all pass with `if (join.getLastAcceptedVersion() > getLastAcceptedVersion()) {`

Reply:
fixed in 8d2fc8cddd6 and acad072945f